### PR TITLE
Queues - Emma Philippides- Ride-Share-Two

### PR DESCRIPTION
**Pull request by eclairedelune**
# Ride Share
Congratulations! You're submitting your assignment!  

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |    One of the hardest decision I had to make was to submit an incomplete assignment. I spent hours on end trying to figure out how to fix the errors which came from the RideShare::Rider::self.find, which only led to getting more errors.  I  made my final decision after having other ladies look at the errors and they could not figure it out either.  |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  I really liked the design diagram and how it made understanding the communication between the classes clearer.  |
| Describe a _nominal test_ that you wrote for this assignment. |  making sure that the self.all Trip method returned an array.  |
| Describe an _edge case test_ that you wrote for this assignment. | testing for a nonexistent trip id and making sure it returned nil   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I wrote the pseudocode on paper while I was working on the design diagram. I felt that it simplified the process, so I will be using more in the future.  |

### CODE REVIEW

File: specs/driver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    it "Returns an existing driver" do
+      existing_driver = RideShare::Driver.find(86)
+      existing_driver.must_be_kind_of RideShare::Driver
+    end
+
+    it "Can find the first driver from the CSV" do
+      first_driver = RideShare::Driver.find(1)
+      first_driver.must_be_instance_of RideShare::Driver
+      first_driver.must_equal 1
```
Comment (droberts-sea) [inline specs/driver_spec.rb:85]: You've just checked that `first_driver` is an instance of `Driver`, so I'm not sure that `first_driver.must_equal 1` here makes sense. Did you mean to check the driver's ID instead?

File: specs/driver_spec.rb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    end
+
+  end
+
+  describe "average_rating" do
+
+    it "returns the average rating of the driver" do
+      average_rating = [3, 5, 4, 3, 2, 5, 2, 4]/average_rating.length
+      RideShare::Trip.average_rating.must_equal 28.0 / 8
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:112]: What happens when you try and find the average rating for a driver with no trips?

File: lib/rider.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        @riders << self.new({rider_id:line[0].to_i, name:line[1], vin:line[2]})
+      end
+      return @riders
+    end
+  end
+
+  def self.find(rider_id)
+    rider_lookup = RiderShare::Rider.all
+    rider_lookup.find { |rider| rider.rider_id == rider_id }
+    puts "invalid rider id" if rider_lookup == nil
```
Comment (droberts-sea) [inline lib/rider.rb:26]: Probably don't need to `puts` here - returning `nil` is usually sufficient to let the caller know that nothing was found.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+  describe "trip.find(driver_id)" do
+
+    it "returns trips using the driver's ID" do
+      driver_trips[0].trip_id.must_equal(55)
+    end
+
+    it "returns an array of all the trips of a driver" do
+      driver_trips.length.must_equal 7
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:103]: A test case I'd like to see: What happens when you call `find_trips` on a driver without any trips? What should you get back?

File: specs/rider_spec.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  describe "RideShare::Rider.find_trips" do
+    let(:rider_trips) { RideShare::Trip.rider_id(54).find_trips }
+    it "returns array of trips associated with the rider's ID" do
+      rider_trips.must_be_instance_of Array
+      rider_trips.length.must_equal 2
+      my_rider.find_trips.each do |trip|
+        trip.must_be_instance_of RideShare::Trip
+      end
+    end
```
Comment (droberts-sea) [inline specs/rider_spec.rb:79]: What does find_trips do if the rider hasn't taken any trips?

File: specs/rider_spec.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    it "Can find the first rider from the CSV" do
+
+      my_rider.must_be_instance_of RideShare::Rider
+      my_rider.rider_id.must_equal 1
+    end
+
+    it "Can find the last rider from the CSV" do
+      my_rider.must_be_instance_of RideShare::Rider
+      my_rider.rider_id.must_equal 300
+    end
```
Comment (droberts-sea) [inline specs/rider_spec.rb:68]: What happens if you call `find` with an invalid ID, or an ID that doesn't correspond to a driver?

File: lib/rider.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    return rider_lookup
+  end
+
+  def find_trips(id)
+    trips = RideShare::Trip.all
```
Comment (droberts-sea) [inline lib/rider.rb:31]: This code should probably live in `Trip.get_trips_by_rider_id`, in which case this function would just call that function.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   yes	|
|   Answer comprehension questions	|   yes	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   missing some - see inline comments |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   see inline comments |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | see inline comment |
|   Created a method that uses the internal trips method to retrieve the associated drivers | no |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |  yes  |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   no |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   no |
|   Created a method to retrieve all trips by driver id |    no |
|   Created a method to retrieve all trips by rider id |     no |
|   **Overall**	|   	|

This is a really good start, there's just a few things still missing. As we talked about in our 1-on-1, it seems like the main points to improve on are debugging and the speed of going from an idea in English to some lines of code.
